src/reanimated2/hook/useScrollViewOffset.ts

```diff
@@ -15,22 +15,69 @@ import { isWeb } from '../PlatformChecker';
 
 const IS_WEB = isWeb();
 
-const scrollEventNames = [
-  'onScroll',
-  'onScrollBeginDrag',
-  'onScrollEndDrag',
-  'onMomentumScrollBegin',
-  'onMomentumScrollEnd',
-];
-
 /**
  * Lets you synchronously get the current offset of a `ScrollView`.
  *
  * @param animatedRef - An [animated ref](https://docs.swmansion.com/react-native-reanimated/docs/core/useAnimatedRef) attached to an Animated.ScrollView component.
  * @returns A shared value which holds the current offset of the `ScrollView`.
  * @see https://docs.swmansion.com/react-native-reanimated/docs/scroll/useScrollViewOffset
  */
-export function useScrollViewOffset(
+export const useScrollViewOffset = IS_WEB
+  ? useScrollViewOffsetJS
+  : useScrollViewOffsetNative;
+
+function useScrollViewOffsetJS(
+  animatedRef: AnimatedRef<AnimatedScrollView>,
+  initialRef?: SharedValue<number>
+): SharedValue<number> {
+  const offsetRef = useRef(
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    initialRef !== undefined ? initialRef : useSharedValue(0)
+  );
+  const scrollRef = useRef<AnimatedScrollView | null>(null);
+
+  const eventHandler = useCallback(() => {
+    'worklet';
+    const element = animatedRef.current as unknown as HTMLElement;
+    // scrollLeft is the X axis scrolled offset, works properly also with RTL layout
+    offsetRef.current.value =
+      element.scrollLeft === 0 ? element.scrollTop : element.scrollLeft;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [animatedRef, animatedRef.current]);
+
+  useEffect(() => {
+    // We need to make sure that listener for old animatedRef value is removed
+    if (scrollRef.current !== null) {
+      (scrollRef.current as unknown as HTMLElement).removeEventListener(
+        'scroll',
+        eventHandler
+      );
+    }
```

Comment:
Why do we need to call `removeEventListener` here as well instead of only in `useEffect` cleanup function?

Reply:
@tomekzaw because the animatedRef reference might change to another scroll and we then want to clean the previous listener (see example in code section of PR description)

Reply:
Can we just assign `const element = animatedRef.current` first (when calling `addEventListener`) and then call `removeEventListener` on `element`?

Reply:
That's exactly what `scrollRef` does here

Reply:
To be precise: we store a ref to the affected component in `scrollRef` each time `useEffect` runs. Since its dependencies are `animatedRef`, `animatedRef.current` and `eventHandler` (which also regenerates only when `animatedRef` and `animatedRef.current` change), it re-runs ONLY when we switch the hook to another component. In such situation, cleanup on `scrollRef` runs first, ensuring there is no listener to a scroll that we don't care about, and then we can safely add new listener to a new component.

Reply:
Oh, I see now. Anyway, is it possible to modify the code so that `removeEventListener` appear only once?

Reply:
I'm afraid I don't see any other way. The `removeEventListener` calls match two different scenarios - the `scrollRef` one makes sure we disconnect from an older scroll component during a switch, while the `return` cleanup runs when the component (or its parent) gets dismounted for good.